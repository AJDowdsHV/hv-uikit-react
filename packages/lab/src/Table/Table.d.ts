import { StandardProps } from "@material-ui/core";

export type HvTableClassKey = "root" | "stickyHeader";

export interface HvTableProps
  extends StandardProps<React.HTMLAttributes<HTMLTableElement>, HvTableClassKey> {
  /**
   * Whether the `HvTable` has a sticky header row
   */
  stickyHeader?: boolean;
}

export default function HvTable(props: HvTableProps): JSX.Element | null;
