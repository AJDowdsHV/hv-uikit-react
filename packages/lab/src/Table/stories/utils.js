import { useState } from "react";
import range from "lodash/range";
import { Random } from "@hv/uikit-react-core";

const rand = new Random();

const formatDate = (date) => date.toISOString().split("T")[0];

const newEntry = (value, i) => {
  const [r1, r2] = range(2).map(() => rand.next());
  const [dateMax, dateMin] = [1640995200000, 1514764800000];
  return {
    id: `${i}`,
    name: `Event ${i}`,
    createdDate: formatDate(new Date(rand.next(dateMax, dateMin))),
    eventType: "Anomaly detection",
    status: i % 2 === 0 ? "Closed" : "Open",
    riskScore: rand.next(100, 10),
    severity: (r1 > 0.66 && "Critical") || (r1 > 0.33 && "Major") || "Minor",
    priority: (r2 > 0.66 && "High") || (r2 > 0.33 && "Medium") || "Low",
  };
};

export const makeData = (len = 10) => Array.from(Array(len), newEntry);

// https://react-table.tanstack.com/docs/api/useTable#column-options
// width is only used if explicitly passed in column.getHeaderProps
export const getColumns = () => [
  { Header: "Title", accessor: "name", width: 260 },
  { Header: "Time", accessor: "createdDate" },
  { Header: "Event Type", accessor: "eventType", width: 200 },
  { Header: "Status", accessor: "status", width: 120 },
  // numeric values should be right-aligned
  {
    Header: "Probability",
    accessor: "riskScore",
    align: "right",
    Cell: ({ value }) => `${value}%`,
  },
  { Header: "Severity", accessor: "severity" },
  { Header: "Priority", accessor: "priority" },
];

export const useToggleIndex = (initialState) => {
  const [index, setIndex] = useState(initialState);

  const toggleState = (idx) => () => {
    setIndex(idx === index ? -1 : idx);
  };

  return [index, toggleState];
};
