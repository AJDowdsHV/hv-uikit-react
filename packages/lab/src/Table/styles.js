const styles = (theme) => ({
  root: {
    width: "100%",

    ...theme.hv.typography.normalText,

    "table&": {
      borderSpacing: 0,
    },

    "& caption": {
      ...theme.typography.normalText,
      padding: theme.spacing("xs"),
      textAlign: "left",
      captionSide: "bottom",
    },
  },
  stickyHeader: {},
  stickyColumns: {
    backgroundColor: theme.hv.palette.atmosphere.atmo2,
  },
});

export default styles;
