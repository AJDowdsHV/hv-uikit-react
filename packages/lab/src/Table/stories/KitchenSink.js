import React, { useState, useMemo } from "react";

import { Ban, Duplicate, Delete, Lock, Preview } from "@hv/uikit-react-icons";
import { HvCheckBox, HvEmptyState } from "@hv/uikit-react-core";
import { useTable, useRowSelect, usePagination, useSortBy } from "react-table";

import {
  HvTable,
  HvTableBody,
  HvTableCell,
  HvTableContainer,
  HvTableHead,
  HvTablePagination,
  HvTableRow,
  HvTableBulkActions,
  HvTableHeader,
} from "../..";

import { makeData, getColumns } from "./utils";

const KitchenSink = () => {
  const [data, setData] = useState(makeData(64));
  const columns = useMemo(
    () => [
      {
        id: "selection",
        padding: "checkbox",
        /* eslint-disable-next-line react/prop-types */
        Cell: ({ row }) => <HvCheckBox {...row.getToggleRowSelectedProps()} />,
      },
      ...getColumns().map((col) => ({ ...col, isSortable: true })),
    ],
    []
  );

  const EmptyRow = () => (
    <HvTableRow>
      <HvTableCell colSpan="100%" style={{ height: 100 }}>
        <HvEmptyState message="No data to display" icon={<Ban role="presentation" />} />
      </HvTableCell>
    </HvTableRow>
  );

  const instance = useTable(
    { columns, data, autoResetSelectedRows: false },
    useSortBy,
    usePagination,
    useRowSelect
  );
  const {
    getTableProps,
    getTableBodyProps,
    prepareRow,
    headers,
    page,
    selectedFlatRows,
  } = instance;

  const handleAction = (evt, id, action) => {
    const selected = selectedFlatRows.map((el) => el.original);
    // eslint-disable-next-line no-console
    console.log(id, action);

    switch (action.id) {
      case "duplicate": {
        const newEls = selected.map((el) => ({
          ...el,
          id: `${el.id}-copy`,
          name: `${el.name}-copy`,
        }));
        setData([...data, ...newEls]);
        break;
      }
      case "delete": {
        const selectedIds = selected.map((el) => el.id);
        setData(data.filter((el) => !selectedIds.includes(el.id)));
        break;
      }
      case "lock":
      case "preview":
      default:
        break;
    }
  };

  const rowRenderer = (pages) => {
    return pages.map((row, index) => {
      prepareRow(row);
      return (
        <HvTableRow
          hover
          key={row.Header}
          selected={row.isSelected}
          {...row.getRowProps({
            "aria-rowindex": index,
          })}
        >
          {row.cells.map((cell) => (
            <HvTableCell key={cell.Header} rtCol={cell.column} {...cell.getCellProps()}>
              {cell.render("Cell")}
            </HvTableCell>
          ))}
        </HvTableRow>
      );
    });
  };
  return (
    <>
      <HvTableBulkActions
        rtInstance={instance}
        maxVisibleActions={1}
        actionsCallback={handleAction}
        actions={[
          { id: "delete", label: "Delete", icon: <Delete /> },
          { id: "duplicate", label: "Duplicate", icon: <Duplicate /> },
          { id: "lock", label: "Lock", icon: <Lock /> },
          { id: "preview", label: "Preview", icon: <Preview /> },
        ]}
      />

      <HvTableContainer>
        <HvTable
          {...getTableProps({
            "aria-rowcount": data.length,
            caption: "Table Caption",
          })}
        >
          <HvTableHead>
            <HvTableRow>
              {headers.map((col) => (
                <HvTableHeader
                  key={col.Header}
                  rtCol={col}
                  {...col.getHeaderProps(col.getSortByToggleProps())}
                >
                  {col.render("Header")}
                </HvTableHeader>
              ))}
            </HvTableRow>
          </HvTableHead>
          <HvTableBody {...getTableBodyProps()}>
            {page.length === 0 ? <EmptyRow /> : rowRenderer(page)}
          </HvTableBody>
        </HvTable>
      </HvTableContainer>
      <HvTablePagination rtInstance={instance} />
    </>
  );
};

export default KitchenSink;
