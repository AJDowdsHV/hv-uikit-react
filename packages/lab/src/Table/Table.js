import React, { forwardRef, useMemo } from "react";
import PropTypes from "prop-types";
import clsx from "clsx";

import { withStyles } from "@material-ui/core";

import TableContext from "./TableContext";
import styles from "./styles";

/**
 * A Table gathers relational data, it displays values arranged to allow quick numerical analysis
 * like comparison and sorting.
 *
 * HvTable component offers a set of HTML-equivalent elements, styled to Design System's specification,
 * for building tables.
 *
 * For better data handling and advanced features we recommend the use of a collection of utility hooks.
 * See the <a href="?id=lab-table-hooks--main&viewMode=docs" target="_self">Table Hooks documentation</a> for more details.
 *
 * PLEASE NOTE: This Table implementation is still a WIP. There might be breaking changes.
 */
const HvTable = forwardRef(function HvTable(props, ref) {
  const { classes, className, stickyHeader = false, ...others } = props;

  const tableContext = useMemo(() => ({ stickyHeader }), [stickyHeader]);

  return (
    <TableContext.Provider value={tableContext}>
      <table
        ref={ref}
        className={clsx(classes.root, className, { [classes.stickyHeader]: stickyHeader })}
        {...others}
      />
    </TableContext.Provider>
  );
});

HvTable.propTypes = {
  /**
   * Class names to be applied.
   */
  className: PropTypes.string,
  /**
   * Content to be rendered
   */
  children: PropTypes.node.isRequired,
  /**
   * Whether the `HvTable` has a sticky header row
   */
  stickyHeader: PropTypes.bool,
  /**
   * A Jss Object used to override or extend the styles applied.
   */
  classes: PropTypes.shape({
    /**
     * Styles applied to the component root class.
     */
    root: PropTypes.string,
    /**
     * Styles applied to the component root class when it has a sticky header.
     */
    stickyHeader: PropTypes.string,
  }).isRequired,
};

export default withStyles(styles, { name: "HvTable" })(HvTable);
