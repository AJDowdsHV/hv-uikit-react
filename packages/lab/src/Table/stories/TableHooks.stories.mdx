import { Meta, Story, Preview } from "@storybook/addon-docs/blocks";

import LinkTo from "@storybook/addon-links/react";

import { Usage } from "storybook-root/blocks/Usage";

import { HvTable } from "../..";

import * as stories from "./TableHooks.stories.js";

<Meta title="Lab/Table Hooks" />

# Table hooks

[React Table](https://react-table.tanstack.com) is an "headless" UI utility library that allows to build data tables while retaining control over markup and styles.

It consists of a collection of custom React hooks. These hooks are lightweight, composable and extensible.
The UI Kit library provides some custom hooks that ease the integration with the <LinkTo kind="Lab/Table" story="Main" className="sbdocs-a">Table component</LinkTo>.

Together is easier to build powerful experiences and extend the features offered.

<Usage>{'import { useTable } from "react-table";'}</Usage>

<Preview withToolbar>
  <Story inline story={stories.Main} />
</Preview>

## Pagination

Data can be paginated. Optionally, empty lines can be added to the last page to avoid changes in height.

Check [React Table's `usePagination` documentation](https://react-table.tanstack.com/docs/api/usePagination) for configuration details.

<Preview withToolbar>
  <Story inline story={stories.Pagination} />
</Preview>

## Selection and bulk actions

Rows can be selected individually or in bulk.

Check [React Table's `useRowSelect` documentation](https://react-table.tanstack.com/docs/api/useRowSelect) for configuration details.

<Preview withToolbar>
  <Story inline story={stories.BulkActions} />
</Preview>

## Sortable

Tables can have the capability of sorting the data by column.
Clicking a sortable column header toggles between ascending or descending. Multiple sorting criteria is enabled through holding shift while clicking.

Check [React Table's `useSortBy` documentation](https://react-table.tanstack.com/docs/api/useSortBy) for configuration details.

<Preview withToolbar>
  <Story inline story={stories.Sortable} />
</Preview>

## Expandable

Tables rows can have the capability of being expanded to show more details.

Check [React Table's `useExpanded` documentation](https://react-table.tanstack.com/docs/api/useExpanded) for configuration details.

<Preview withToolbar>
  <Story inline story={stories.Expandable} />
</Preview>

## Sticky Headers And Columns

The table headers can be fixed at the top, allowing it to keep always visible when table has vertical scrolling enabled.

It is also possible to stick columns, either on the left or on the right, making them always visible when scrolling horizontally.

This is implemented the `useHvTableSticky` custom React Table hook provided by the UI Kit library.

When using sticky columns, all column widths will be fixed and a non-table-element layout is applied.

<Usage>{'import { useHvTableSticky } from "@hv/uikit-react-core";'}</Usage>

### Configuration

#### Table Options

The following options are supported via the main options object passed to `useTable(options)`:

- `stickyHeader: Bool`
  - Enables the sticky header

#### Column Options

The following options are supported on any `Column` object passed to the `columns` options in `useTable()`:

- `sticky: "left" | "right"`
  - Sticks this column to the defined side. Columns will be re-orderer if needed.
- `width: Int`
  - Defaults to 150
  - Specifies the width for the column

### Usage

The `getTableProps`, `getTableHeadProps`, `getHeaderGroupProps`, `getHeaderProps`, `getTableBodyProps`, `getRowProps` and `getCellProps` prop getters must be called and the properties injected into the corresponding elements.

<Preview withToolbar>
  <Story inline story={stories.StickyHeadersAndColumns} />
</Preview>

## Empty Cells

The absence of data within an individual cell in the table should be displayed as a `â€”` (_em dash_), according to Design System guidelines.

This must be handled by the developer in the respective cell renderer, as it is an guideline and not a strict rule.
Different use cases may require different solutions, like `N/A` or `0`. Empty cells should be avoided.

<Preview withToolbar>
  <Story inline story={stories.EmptyCells} />
</Preview>

# Stuff still to be documented

## Bulk Actions Manual

<Preview withToolbar>
  <Story inline story={stories.BulkActionsManual} />
</Preview>

## Server Side

<Preview withToolbar>
  <Story inline story={stories.ServerSide} />
</Preview>
