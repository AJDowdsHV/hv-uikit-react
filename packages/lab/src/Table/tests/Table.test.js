/* eslint-env jest */
import React from "react";
import range from "lodash/range";
import { render, within } from "testing-utils";
import userEvent from "@testing-library/user-event";

import { Main, EmptyCells, NonTableLayout } from "../stories/Table.stories";

import { Pagination, Sortable, BulkActions } from "../stories/TableHooks.stories";

import {
  HvTable,
  HvTableBody,
  HvTableCell,
  HvTableContainer,
  HvTableHead,
  HvTableHeader,
  HvTableRow,
} from "../..";

describe("Table", () => {
  describe("Main Story", () => {
    it("should match snapshot", () => {
      const { container } = render(<Main />);
      expect(container).toBeDefined();
    });

    it("should render the table elements", () => {
      const { getByRole, getAllByRole } = render(<Main />);

      expect(getByRole("table")).toBeInTheDocument();

      expect(getAllByRole("cell").length).toBeGreaterThan(0);
      expect(getAllByRole("row").length).toBeGreaterThan(0);
    });
  });

  describe("non-table elements support", () => {
    it("defaults to table elements", () => {
      const { getByRole, getAllByRole } = render(<Main />);

      const table = getByRole("table");
      expect(table).toBeInTheDocument();

      expect(table.nodeName).toEqual("TABLE");
      expect(table).not.toHaveAttribute("role");

      const rowgroups = getAllByRole("rowgroup");
      expect(rowgroups.length).toBe(2);

      expect(rowgroups[0].nodeName).toEqual("THEAD");
      expect(rowgroups[0]).not.toHaveAttribute("role");

      expect(rowgroups[1].nodeName).toEqual("TBODY");
      expect(rowgroups[1]).not.toHaveAttribute("role");

      const headerRows = within(rowgroups[0]).getAllByRole("row");
      expect(headerRows.length).toBe(1);

      expect(headerRows[0].nodeName).toEqual("TR");
      expect(headerRows[0]).not.toHaveAttribute("role");

      const headerCells = within(headerRows[0]).getAllByRole("columnheader");
      expect(headerCells.length).toBe(7);

      expect(headerCells[0].nodeName).toEqual("TH");
      expect(headerCells[0]).not.toHaveAttribute("role");

      const bodyRows = within(rowgroups[1]).getAllByRole("row");
      expect(bodyRows.length).toBe(6);

      expect(bodyRows[0].nodeName).toEqual("TR");
      expect(bodyRows[0]).not.toHaveAttribute("role");

      const bodyCells = within(bodyRows[0]).getAllByRole("cell");
      expect(bodyCells.length).toBe(7);

      expect(bodyCells[0].nodeName).toEqual("TD");
      expect(bodyCells[0]).not.toHaveAttribute("role");
    });

    it("can render a different component and sets roles", () => {
      const { getByRole, getAllByRole } = render(<NonTableLayout />);

      const table = getByRole("table");
      expect(table).toBeInTheDocument();

      expect(table.nodeName).toEqual("DIV");
      expect(table).toHaveAttribute("role", "table");

      const rowgroups = getAllByRole("rowgroup");
      expect(rowgroups.length).toBe(2);

      expect(rowgroups[0].nodeName).toEqual("DIV");
      expect(rowgroups[0]).toHaveAttribute("role", "rowgroup");

      expect(rowgroups[1].nodeName).toEqual("DIV");
      expect(rowgroups[1]).toHaveAttribute("role", "rowgroup");

      const headerRows = within(rowgroups[0]).getAllByRole("row");
      expect(headerRows.length).toBe(1);

      expect(headerRows[0].nodeName).toEqual("DIV");
      expect(headerRows[0]).toHaveAttribute("role", "row");

      const headerCells = within(headerRows[0]).getAllByRole("columnheader");
      expect(headerCells.length).toBe(7);

      expect(headerCells[0].nodeName).toEqual("DIV");
      expect(headerCells[0]).toHaveAttribute("role", "columnheader");

      const bodyRows = within(rowgroups[1]).getAllByRole("row");
      expect(bodyRows.length).toBe(6);

      expect(bodyRows[0].nodeName).toEqual("DIV");
      expect(bodyRows[0]).toHaveAttribute("role", "row");

      const bodyCells = within(bodyRows[0]).getAllByRole("cell");
      expect(bodyCells.length).toBe(7);

      expect(bodyCells[0].nodeName).toEqual("DIV");
      expect(bodyCells[0]).toHaveAttribute("role", "cell");
    });
  });

  describe("Simple Table", () => {
    const NUM_ROWS = 6;
    const NUM_COLS = 3;

    const SimpleTable = () => (
      <HvTableContainer>
        <HvTable>
          <HvTableHead>
            <HvTableRow>
              {range(NUM_COLS).map((id) => (
                <HvTableHeader key={id}>{`Sample Header ${id}`}</HvTableHeader>
              ))}
            </HvTableRow>
          </HvTableHead>
          <HvTableBody>
            {range(NUM_ROWS).map((id) => (
              <HvTableRow key={id}>
                {range(NUM_COLS).map((id2) => (
                  <HvTableCell key={id2}>{`Sample Cell ${id2}`}</HvTableCell>
                ))}
              </HvTableRow>
            ))}
          </HvTableBody>
        </HvTable>
      </HvTableContainer>
    );

    it("should be defined", () => {
      const { container } = render(<SimpleTable />);

      expect(container).toBeDefined();
      expect(container).toMatchSnapshot();
    });

    it("should render the rows and cells", () => {
      const { getAllByRole } = render(<SimpleTable />);

      expect(getAllByRole("rowgroup").length).toBe(2); // thead & tbody
      expect(getAllByRole("row").length).toBe(NUM_ROWS + 1);
      expect(getAllByRole("columnheader").length).toBe(NUM_COLS);
      expect(getAllByRole("cell").length).toBe(NUM_COLS * NUM_ROWS);
    });
  });

  describe("Header Only Table", () => {
    const NUM_COLS = 6;

    const HeadersOnlyTable = () => (
      <HvTableContainer>
        <HvTable>
          <HvTableHead>
            <HvTableRow>
              {range(NUM_COLS).map((id) => (
                <HvTableHeader key={id}>{`Header ${id}`}</HvTableHeader>
              ))}
            </HvTableRow>
          </HvTableHead>
          <HvTableBody />
        </HvTable>
      </HvTableContainer>
    );

    it("should render single row and its cells", () => {
      const { getAllByRole } = render(<HeadersOnlyTable />);

      expect(getAllByRole("row").length).toBe(1);
      expect(getAllByRole("columnheader").length).toBe(NUM_COLS);
    });
  });

  describe("EmptyCells Story", () => {
    it("should be defined", () => {
      const { container } = render(<EmptyCells />);
      expect(container).toBeDefined();
    });

    it("should contain the em-dashes", () => {
      const { getAllByText } = render(<EmptyCells />);

      expect(getAllByText("â€”").length).toBe(3);
    });
  });

  describe("Pagination Story", () => {
    it("should be defined", () => {
      const { container } = render(<Pagination />);
      expect(container).toBeDefined();
    });

    it("should render the table elements", () => {
      const { getByRole, getAllByRole } = render(<Pagination />);

      expect(getByRole("table")).toBeInTheDocument();

      expect(getAllByRole("cell").length).toBeGreaterThan(0);
      expect(getAllByRole("row").length).toBeGreaterThan(0);
    });

    it("should contain the correct page elements", () => {
      const { getByLabelText, queryByText, getAllByRole } = render(<Pagination />);

      const [fistPage, previousPage, nextPage, lastPage] = [
        "First Page",
        "Previous Page",
        "Next Page",
        "Last Page",
      ].map(getByLabelText);

      expect(getAllByRole("row").length).toBe(11);
      expect(queryByText("Event 1")).toBeInTheDocument();
      expect(queryByText("Event 11")).not.toBeInTheDocument();

      userEvent.click(fistPage);
      expect(getAllByRole("row").length).toBe(11);
      expect(queryByText("Event 1")).toBeInTheDocument();
      expect(queryByText("Event 11")).not.toBeInTheDocument();

      userEvent.click(previousPage);
      expect(getAllByRole("row").length).toBe(11);
      expect(queryByText("Event 1")).toBeInTheDocument();
      expect(queryByText("Event 11")).not.toBeInTheDocument();

      userEvent.click(nextPage);
      expect(getAllByRole("row").length).toBe(11);
      expect(queryByText("Event 1")).not.toBeInTheDocument();
      expect(queryByText("Event 11")).toBeInTheDocument();

      userEvent.click(lastPage);
      expect(getAllByRole("row").length).toBe(11);
      expect(queryByText("Event 1")).not.toBeInTheDocument();
      expect(queryByText("Event 30")).not.toBeInTheDocument();
      expect(queryByText("Event 31")).toBeInTheDocument();
    });
  });

  describe("Sortable Story", () => {
    it("should be defined", () => {
      const { container } = render(<Sortable />);
      expect(container).toBeDefined();
    });

    it("should render the table elements", () => {
      const { getByRole, getAllByRole } = render(<Sortable />);

      expect(getByRole("table")).toBeInTheDocument();

      expect(getAllByRole("cell").length).toBeGreaterThan(0);
      expect(getAllByRole("row").length).toBeGreaterThan(0);
    });

    it("should single-sort as expected", () => {
      const { getByText, queryAllByText } = render(<Sortable />);
      const severityButton = getByText("Severity");

      const getEvent = (i) => queryAllByText(/^Event \d+$/g)[i];

      // Default sorting
      expect(getEvent(0)).toHaveTextContent("Event 1");

      // Asc sorting
      userEvent.click(severityButton);
      expect(getEvent(0)).toHaveTextContent("Event 4");

      // Desc sorting
      userEvent.click(severityButton);
      expect(getEvent(0)).toHaveTextContent("Event 5");

      // Back to default sorting
      userEvent.click(severityButton);
      expect(getEvent(0)).toHaveTextContent("Event 1");
    });
  });

  describe("BulkActions Story", () => {
    it("should be defined", () => {
      const { container } = render(<BulkActions />);
      expect(container).toBeDefined();
    });

    it("should render the table elements", () => {
      const { getByRole, getAllByRole } = render(<BulkActions />);

      expect(getByRole("table")).toBeInTheDocument();

      expect(getAllByRole("cell").length).toBeGreaterThan(0);
      expect(getAllByRole("row").length).toBeGreaterThan(0);
    });

    it("should select rows accordingly expected", () => {
      const { getByLabelText, getAllByRole, queryAllByRole } = render(<BulkActions />);
      const bulkCheckbox = getAllByRole("checkbox")[0];
      const firstCheckbox = getAllByRole("checkbox")[1];
      const secondCheckbox = getAllByRole("checkbox")[2];

      expect(getByLabelText("All (64)")).toBeInTheDocument();
      expect(bulkCheckbox).toBeInTheDocument();
      expect(firstCheckbox).toBeInTheDocument();
      expect(secondCheckbox).toBeInTheDocument();
      expect(queryAllByRole("checkbox", { checked: true }).length).toBe(0);
      expect(queryAllByRole("checkbox", { checked: false }).length).toBe(10 + 1);

      // Click a checkbox
      userEvent.click(firstCheckbox);
      expect(getByLabelText("1 / 64")).toBeInTheDocument();
      expect(queryAllByRole("checkbox", { checked: true }).length).toBe(1 + 1);

      // Click another checkbox
      userEvent.click(secondCheckbox);
      expect(getByLabelText("2 / 64")).toBeInTheDocument();
      expect(queryAllByRole("checkbox", { checked: true }).length).toBe(2 + 1);

      // Click Bulk Actions - should deselect all in page when any selected
      userEvent.click(bulkCheckbox);
      expect(getByLabelText("All (64)")).toBeInTheDocument();
      expect(queryAllByRole("checkbox", { checked: true }).length).toBe(0);

      // Click Bulk Actions - should select all in page when none selected
      userEvent.click(bulkCheckbox);
      expect(getByLabelText("10 / 64")).toBeInTheDocument();
      expect(queryAllByRole("checkbox", { checked: true }).length).toBe(10 + 1);
    });
  });

  describe("Table line click", () => {
    it("executes the provided callback", () => {
      const onClickSpy = jest.fn();

      const TableWithRowClick = () => {
        const NUM_ROWS = 6;
        const NUM_COLS = 3;

        return (
          <HvTableContainer>
            <HvTable>
              <HvTableHead>
                <HvTableRow>
                  {range(NUM_COLS).map((id) => (
                    <HvTableHeader key={id}>{`Sample Header ${id}`}</HvTableHeader>
                  ))}
                </HvTableRow>
              </HvTableHead>
              <HvTableBody>
                {range(NUM_ROWS).map((id) => (
                  <HvTableRow key={id} onClick={(e) => onClickSpy(id, e)}>
                    {range(NUM_COLS).map((id2) => (
                      <HvTableCell key={id2}>{`Sample Cell ${id2}`}</HvTableCell>
                    ))}
                  </HvTableRow>
                ))}
              </HvTableBody>
            </HvTable>
          </HvTableContainer>
        );
      };

      const { getAllByRole } = render(<TableWithRowClick />);

      const rowsInTable = getAllByRole("row");

      userEvent.click(rowsInTable[1]);
      userEvent.click(rowsInTable[3]);

      expect(onClickSpy).toHaveBeenCalledTimes(2);

      expect(onClickSpy).toHaveBeenNthCalledWith(1, 0, expect.anything());
      expect(onClickSpy).toHaveBeenNthCalledWith(2, 2, expect.anything());
    });
  });
});
