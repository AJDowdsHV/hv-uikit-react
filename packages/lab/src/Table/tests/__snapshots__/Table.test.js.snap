// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Simple Table should be defined 1`] = `
<div>
  <div
    class="HvTableContainer-root"
  >
    <table
      class="HvTable-root"
    >
      <thead
        class="HvTableHead-root"
      >
        <tr
          class="HvTableRow-root HvTableRow-head"
        >
          <th
            class="HvTableCell-root HvTableCell-head"
          >
            Sample Header 0
          </th>
          <th
            class="HvTableCell-root HvTableCell-head"
          >
            Sample Header 1
          </th>
          <th
            class="HvTableCell-root HvTableCell-head"
          >
            Sample Header 2
          </th>
        </tr>
      </thead>
      <tbody
        class="HvTableBody-root"
      >
        <tr
          class="HvTableRow-root HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 0
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 1
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 2
          </td>
        </tr>
        <tr
          class="HvTableRow-root HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 0
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 1
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 2
          </td>
        </tr>
        <tr
          class="HvTableRow-root HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 0
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 1
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 2
          </td>
        </tr>
        <tr
          class="HvTableRow-root HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 0
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 1
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 2
          </td>
        </tr>
        <tr
          class="HvTableRow-root HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 0
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 1
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 2
          </td>
        </tr>
        <tr
          class="HvTableRow-root HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 0
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 1
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Sample Cell 2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
